### PR TITLE
Introducing preview version of linchpin api V2 for 1.7.6

### LINKED ISSUE #1057: RFE: Remove context and other functions which are used while initiating linchpin API
**Is your feature request related to a problem? Please describe.**
Currently it is very difficult understand and intiate provisioning of linchpin API. Current request is to simplify the process with less code. 
**Describe the solution you'd like**
Current solution: 
```python
from linchpin import LinchpinAPI
from linchpin.context import LinchpinContext

context = LinchpinContext()
context.setup_logging()
context.load_config()
context.load_global_evars()
context.set_cfg('lp', 'workspace', '.')
context.set_evar('workspace', '.')
context.set_evar('debug_mode', True)
linchpin_api = LinchpinAPI(context)
pindict = {
  "simple": {
    "layout": {
      "inventory_layout": {
        "hosts": {
          "example-node": {
            "count": 1,
            "host_groups": [
              "example"
            ]
          }
        },
        "vars": {
          "hostname": "__IP__",
          "ansible_ssh_private_key_file": "~/.ssh/id_rsa"
        }
      }
    },
    "topology": {
      "topology_name": "simple",
      "resource_groups": [
        {
          "resource_group_name": "os-server-new",
          "resource_definitions": [
            {
              "count": 1,
              "name": "database",
              "image": "CentOS-7-x86_64-GenericCloud-1612",
              "keypair": "ci-factory",
              "role": "os_server",
              "fip_pool": "10.8.240.0",
              "flavor": "m1.small",
              "networks": [
                "QE-test"
              ]
            }
          ],
          "resource_group_type": "openstack",
          "credentials": {
            "filename": "clouds.yaml",
            "profile": "default"
          }
        }
      ]
    }
  }
}

# credentials alternatives: file vs environment variables
linchpin_api.do_action(pindict, action='up')
```
Proposed solution:

- Remove context from LinchpinAPI as requirement and add defaults if needed
- Authentication as code - support file, environment variables AND api direct call to ansible_runner -> VariableManager 
something similar to 
```python
import linchpin 
from linchpin import Pinfile

pf = Pinfile(filename="pathtopinfile.txt" or dict )

pf.set_config(conf="path_to_conf")
pf.set_workspace(workspace="other_than_default_ws")

pf.up()
pf.destroy()
```

### PR DESCRIPTION
**Pull request by samvarankashyap**
$subject 

Summary: 
- includes simplified API restructuring maintaining backward compatibility
- Added example documentation
- Added unit tests for linchpin new API
- updated autodocs module in sphinx

fixes #1057

### CODE REVIEW

File: linchpin/__init__.py [hunk 1]
```diff
@@ -723,6 +723,74 @@ def do_validation(self, provision_data, old_schema=False):
...
+                        t_data = targets[name]["inputs"]["topology_data"]
+                        c_data = {}
+                        if "cfgs" in list(targets[name].keys()):
+                            c_data = targets[name]["cfgs"]["user"]
+                        # i_path = targets[name]["outputs"]["inventory_path"][0]
+                        layout = lt_data["inventory_layout"]
+                        """
+                        # check whether inventory_file is mentioned in layout
+
+                        #if not os.path.exists(os.path.dirname(i_path)):
```
Comment (abraverm) [inline linchpin/__init__.py]: Do we need this commented out code?

File: linchpin/tests/test_init_pass.py [hunk 2]
```diff
@@ -28,6 +33,9 @@ def test_api_create():
     assert_equal(isinstance(lpa, LinchpinAPI), True)
 
 
+
```
Comment (abraverm) [inline linchpin/tests/test_init_pass.py]: I think you can remove these empty lines

**Comment by samvarankashyap** [comment]
> First, Kudos! this is great and I think we are on the right direction.
> 
> One thing that I would like to see, and we don't have to block this change, is an option to pass credentials with the api:
> 
> ```
> wrks.credentials = linchpin.api.credentials.openstack(name='psi', auth_url=...)
> ```
> 
> Maybe something simlpler like:
> 
> ```
> creds = """
> [psi]
> auth_url = ...
> """
> wksp.load_credentials(creds)
> ```

currently we do have helper function to set credential_path 
```
wksp.set_credential_path() 
```
function to help users to point out to the credential space. 
Given, multiple formats of credentials for each cloud provider. I believe it would take time implementing and testing them. 
I will create a new issue around it and see to it is implemented in 2.0. 
This API implemented in PR is preview version which will be available from 1.7.6 release which is due July 10th.

File: linchpin/__init__.py (hunk 1 above)
Comment (samvarankashyap) [reply linchpin/__init__.py]: Will remove them.

File: linchpin/tests/test_init_pass.py (hunk 2 above)
Comment (samvarankashyap) [reply linchpin/tests/test_init_pass.py]: Will remove them.

**Comment by 14rcole** [comment]
If I provision linchpin with the API will it automatically check for configuration options set in environment variables?

**Comment by samvarankashyap** [comment]
> If I provision linchpin with the API will it automatically check for configuration options set in environment variables?

@14rcole some of them, it should like linchpin.conf is fetched from default locations. 
What configuration options are we talking about ? if its creds_path I don't think so, users should set the credentials_path.

**Comment by 14rcole** [comment]
@samvarankashyap I was referring to things like OS_USERNAME and whatever kinit sets up for beaker

**Comment by samvarankashyap** [comment]
@14rcole The other cloud provider related environment variables detection remains the same since they are dependent on the ansible modules.

**Comment by samvarankashyap** [comment]
@samvarankashyap Add integration test to the API.

**Comment by samvarankashyap** [comment]
[merge]

**Pull request merged**